Fix parse empty df

bkandel opened the pull request:
- [x] closes #14515 
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

This commit fixes a bug where `read_csv` failed when given a file with a multiindex header and empty content. Because pandas reads index names as a separate line following the header lines, the reader looks for the line with index names in it. If the content of the dataframe is empty, the reader will choke. This bug surfaced after https://github.com/pandas-dev/pandas/issues/6618 stopped writing an extra line after multiindex columns, which led to a situation where pandas could write CSV's that it couldn't then read. 

This commit changes that behavior by explicitly checking if the index name row exists, and processing it correctly if it doesn't.

Commit:
BUG: read_csv with empty df

read_csv would fail on files if the number of header lines passed in includes
all the lines in the files. This commit fixes that bug.

Commit:
pep8

jreback commented on doc/source/whatsnew/v0.19.1.txt:
```diff
@@ -57,5 +57,6 @@ Bug Fixes
 - Bug in ``DataFrame.to_json`` where ``lines=True`` and a value contained a ``}`` character (:issue:`14391`)
 - Bug in ``df.groupby`` causing an ``AttributeError`` when grouping a single index frame by a column and the index level (:issue`14327`)
 - Bug in ``df.groupby`` where ``TypeError`` raised when ``pd.Grouper(key=...)`` is passed in a list (:issue:`14334`)
+- Bug in ``pd.read_csv`` where reading files fails if the number of headers is equal to the number of lines in the file (:issue:`14515`)
```
0.20

jreback commented on pandas/io/tests/parser/common.py:
```diff
@@ -606,6 +606,24 @@ def test_multi_index_no_level_names(self):
+"""
+        df = self.read_csv(StringIO(data), header=[0])
+        expected = DataFrame(columns=[('a'), ('b')])
+        tm.assert_frame_equal(df, expected)
```
use the same styling as above
e.g. expected

jreback commented on pandas/parser.pyx:
```diff
@@ -714,7 +714,9 @@ cdef class TextReader:
                 # e.g., if header=3 and file only has 2 lines
-                elif self.parser.lines < hr + 1:
+                if (self.parser.lines < hr + 1
+                    and not isinstance(self.orig_header, list)) or (
```
this is really odd what r u trying to do

bkandel replied:
Counteract the extension to the header added in: https://github.com/pandas-dev/pandas/blob/master/pandas/parser.pyx#L519 The issue is that if the header is passed in as a list, it's extended to enable reading in the index name line (I think that's what it's for if I interpreted that comment correctly) but not accounted for when checking if the file is too short. So this checks if the header has been artificially extended and doesn't complain about the file being too short if the header was extended beyond the length of the file.

Commit:
typo

Commit:
remove unnecessary test

jorisvandenbossche commented:
@bkandel You have several tests that are failing now (I think also the one you added)

Commit:
fix python parser too

bkandel commented:
@jorisvandenbossche Sorry for the delay in fixing this -- there was more complexity here than I realized. I think it's getting there.

Commit:
Modify test.

A test in test_to_csv checked for the presence of exactly the behavior we're
fixing here: A file with 5 lines that asks for a header of length 5 should work
and return an empty dataframe, not error.

Commit:
whatsnew

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -80,3 +80,4 @@ Performance Improvements
 
 Bug Fixes
 ~~~~~~~~~
+- Bug in ``pd.read_csv`` where reading files fails if the number of headers is equal to the number of lines in the file (:issue:`14515`)
```
can move to 0.19.2

jreback commented on pandas/io/tests/parser/common.py:
```diff
@@ -606,6 +606,21 @@ def test_multi_index_no_level_names(self):
+        data = """a,b
+"""
+        df = self.read_csv(StringIO(data), header=[0])
+        expected = DataFrame(columns=[('a',), ('b',)])
```
I would expect these to resolve to an ``Index`` actually (we don't have a single level MI)

```
In [6]: pd.MultiIndex.from_tuples([('a',), ('b',)])
Out[6]: Index(['a', 'b'], dtype='object')
```

bkandel replied:
That's what I thought too, but I got dtype matching errors when I did that. I'll try to figure out what's going on.

jreback commented on pandas/io/tests/parser/common.py:
```diff
@@ -606,6 +606,21 @@ def test_multi_index_no_level_names(self):
+"""
+        df2 = self.read_csv(StringIO(data_multiline), header=[0, 1])
+        expected2 = DataFrame(columns=[('a', 'c'), ('b', 'd')])
+        tm.assert_frame_equal(df2, expected2)
```
then this works nicely

```
In [7]: pd.MultiIndex.from_tuples([('a','c'), ('b', 'd')])
Out[7]: 
MultiIndex(levels=[['a', 'b'], ['c', 'd']],
           labels=[[0, 1], [0, 1]])
```

Commit:
fix errant this_columns

Commit:
move to 0.19.2

Commit:
fix multiindex column parsing

Commit:
lint

Commit:
lint

jreback commented:
@bkandel looks pretty good. ping when all green.

bkandel commented:
@jreback all green. Should be ready for final review now.

bkandel commented:
@jreback just rebased and should be good to go now.

jreback commented:
thanks!
